File: src/main/java/ux/ControlPanel.java
package ux;


import actions.KeyCallback;

public class ControlPanel extends ContainerPanel{

    ControlPanel(Builder builder){
        super(builder);
    }

    public static class Builder extends ContainerPanel.Builder<Builder>{
        public Builder(DrawingInfoSupplier drawingInformer, AlignHorizontal alignHorizontal, AlignVertical vAlign) {
            super(drawingInformer, alignHorizontal, vAlign);
        }

        public Builder setOrientation(Orientation orientation) {
            this.orientation = orientation;
            return this;
        }

        public Builder addControl(String iconName, KeyCallback callback ) {

            Control c = new Control.Builder(
                    this.drawingInformer,
                    callback,
                    iconName,
                    UXThemeManager.getInstance().getControlSize()
            ).build();

            addPanel(c);

            return this;
        }

        public Builder addToggleHighlightControl(String iconName, KeyCallback callback ) {

            Control c = new ToggleHighlightControl.Builder(
                    this.drawingInformer,
                    callback,
                    iconName,
                    UXThemeManager.getInstance().getControlSize()
            ).build();

            addPanel(c);

            return this;
        }

        public Builder addToggleIconControl(String iconName, String toggledIconName, KeyCallback callback ) {

            Control c = new ToggleIconControl.Builder(
                    this.drawingInformer,
                    callback,
                    iconName,
                    toggledIconName,
                    UXThemeManager.getInstance().getControlSize()
            ).build();

            addPanel(c);

            return this;
        }

        @Override
        public ControlPanel build() {
            return new ControlPanel(this);
        }

        @Override
        protected Builder self() {
            return this;
        }
    }

}

File: src/main/kotlin/patterning/ux/panel/ControlPanel.kt
package patterning.ux.panel

import patterning.RunningState
import patterning.actions.KeyCallback
import patterning.ux.Theme
import patterning.ux.informer.DrawingInfoSupplier

class ControlPanel internal constructor(builder: Builder) : ContainerPanel(builder) {
    class Builder(drawingInformer: DrawingInfoSupplier?, hAlign: AlignHorizontal?, vAlign: AlignVertical?) :
        ContainerPanel.Builder<Builder>(drawingInformer, hAlign, vAlign) {
        public override fun setOrientation(orientation: Orientation): Builder {
            this.orientation = orientation
            return this
        }

        fun addControl(iconName: String, callback: KeyCallback): Builder {
            val c = Control.Builder(
                drawingInformer,
                callback,
                iconName,
                Theme.controlSize
            ).build()
            addPanel(c!!)
            return this
        }

        fun addToggleHighlightControl(iconName: String, callback: KeyCallback): Builder {
            val c: Control = ToggleHighlightControl.Builder(
                drawingInformer,
                callback,
                iconName,
                Theme.controlSize
            ).build()
            addPanel(c)
            return this
        }

        fun addPlayPauseControl(
            playIconName: String,
            pauseIconName: String,
            callback: KeyCallback,
            modeChangeCallback: KeyCallback,
            getRunningState: () -> RunningState
        ): Builder {
            val c: Control = PlayPauseControl.Builder(
                drawingInformer,
                callback,
                modeChangeCallback,
                playIconName,
                pauseIconName,
                Theme.controlSize,
                getRunningState
            ).build()
            addPanel(c)
            return this
        }

        override fun build(): ControlPanel {
            return ControlPanel(this)
        }

        override fun self(): Builder {
            return this
        }
    }
}